### PR TITLE
Fix Date precision error

### LINKED ISSUE #31: Date encoding and decoding does not compare equal
### Describe the bug

I've noticed that if you insert a date into an SQLite database then read it back out, the date's do not compare equal. It appears to be some precision loss due to `Date`s internal use of `timeIntervalSinceReferenceDate` as it does not happen all the time. However, it can be reliably reproduced using Xcode's "run Repeatedly" feature (see reproducer below).

### To Reproduce

Modify the `testTimestampStorage` test to check for date equality like so and run the test, possibly multiple times using Xcode's "run Repeatedly" feature:

```swift
func testTimestampStorage() throws {
    let conn = try SQLiteConnection.open(storage: .memory, threadPool: self.threadPool, on: self.eventLoop).wait()
    defer { try! conn.close().wait() }

    let date = Date()
    let rows = try conn.query("SELECT ? as date", [date.sqliteData!]).wait()
    XCTAssertEqual(rows[0].column("date"), .float(date.timeIntervalSince1970))
    XCTAssertEqual(
        Date(sqliteData: rows[0].column("date")!),
        date
    )
    XCTAssertEqual(
        Date(sqliteData: rows[0].column("date")!)!.timeIntervalSinceReferenceDate,
        date.timeIntervalSinceReferenceDate
    )
    XCTAssertEqual(Date(sqliteData: rows[0].column("date")!)?.description, date.description)
}
```

You will see the following errors:

Screen Shot 2022-11-03 at 4 18 37 PM

The date equality check does not yield useful information but the `timeIntervalSinceReferenceDate` check reveals the underlying issue with precision. Date internally uses `timeIntervalSinceReferenceDate` for equality checking, see: https://github.com/apple/swift-corelibs-foundation/blob/swift-5.7.1-RELEASE/Sources/Foundation/Date.swift#L156

### Expected behavior

When creating a new `Date` in swift then inserting and reading that date to/from the database I would expect that the read date to compare equal to the original.

### Environment

* Vapor Framework version: N/A, reproduces in SQLiteNIO, commit cb114dc718b302070bf95733109375f99b5ab35c (main)
* Vapor Toolbox version: N/A, reproduces in SQLiteNIO
* OS version: macOS 12.6
* Xcode version: 14.0.1

### Additional context

I noticed this originally using plain `fluent-kit` (no Vapor) and traced it down to the SQLiteNIO package but did see it manifest through each level of the SQLite driver (i.e. FluentSQLiteDriver, SQLiteKit, and SQLiteNIO)

### PR DESCRIPTION
**Pull request by Austinpayne**
* Fixes #31

Due to the nature of the issue, this doesn't reproduce every time. But it does reliably reproduce after a couple repeated runs. The fix here was run 100k times to confirm the fix.

### CODE REVIEW

File: Tests/SQLiteNIOTests/SQLiteNIOTests.swift [hunk 1]
```diff
@@ -26,6 +26,8 @@ final class SQLiteNIOTests: XCTestCase {
         let rows = try conn.query("SELECT ? as date", [date.sqliteData!]).wait()
         XCTAssertEqual(rows[0].column("date"), .float(date.timeIntervalSince1970))
         XCTAssertEqual(Date(sqliteData: rows[0].column("date")!)?.description, date.description)
+        XCTAssertEqual(Date(sqliteData: rows[0].column("date")!), date)
+        XCTAssertEqual(Date(sqliteData: rows[0].column("date")!)?.timeIntervalSinceReferenceDate, date.timeIntervalSinceReferenceDate)
```
Comment (gwynne) [inline Tests/SQLiteNIOTests/SQLiteNIOTests.swift:30]: Instead of relying on repeated test runs to demonstrate the correctness of the behavior, can you:

1. Leave these additions to this test in place, these assertions are good to have anyhow.
2. Run the test against a version _without_ your fix until you come across a particular value which triggers the failure (or if you already know a specific value that does so, just grab that, of course 😅). If you by chance happen to find or already have more than one such value, even better!
3. Add a new test which uses the affected value(s) specifically, rather than `Date()`, so failure of the old code and success of your fixed version are visible and deterministic.
4. Finally, add another new test which verifies that timestamps are always clamped to µs precision - start with a higher precision, roundtrip through `SQLiteDataConvertible` (you don't need to do it via a query, really)  and assert that only microseconds are retained, including being rounded up or down as appropriate.

File: Sources/SQLiteNIO/SQLiteDataConvertible.swift [hunk 2]
```diff
@@ -126,7 +126,10 @@ extension Date: SQLiteDataConvertible {
...
             return nil
         }
-        self.init(timeIntervalSince1970: value)
+        // Round to microseconds to avoid nanosecond precision error causing Dates to fail equality
+        let valueSinceReferenceDate = value - Date.timeIntervalBetween1970AndReferenceDate
+        let microsecondsSinceReference = round(valueSinceReferenceDate * 1e6) / 1e6
+        self.init(timeIntervalSinceReferenceDate: microsecondsSinceReference)
     }
 
     public var sqliteData: SQLiteData? {
```
Comment (gwynne) [inline Sources/SQLiteNIO/SQLiteDataConvertible.swift:135]: Shouldn't we be rounding dates going in both directions? If we're effectively only able to load 6 digits worth of precision, ostensibly we should be unable to _store_ more than six digits either. It's true that PostgresNIO's old API (`PostgresDataConvertible`) only bothers for decoding, but @fabianfett's considerably improved `PostgresCodable` API explicitly deals in microseconds in both cases: https://github.com/vapor/postgres-nio/blob/main/Sources/PostgresNIO/New/Data/Date%2BPostgresCodable.swift

Don't forget to update the unit tests to check for correctness on both code paths 🙃 

(P.S.: Yes, Fabian, I know this is the kind of thing that makes you want to deprecate the old ones 🤣)

**Review by gwynne — CHANGES_REQUESTED** [review]
This is a great start! A couple things still need to be added, I commented inline with details

File: Tests/SQLiteNIOTests/SQLiteNIOTests.swift (hunk 1 above)
Comment (Austinpayne) [reply Tests/SQLiteNIOTests/SQLiteNIOTests.swift:30]: Awesome suggestion! I have added these tests.

File: Sources/SQLiteNIO/SQLiteDataConvertible.swift (hunk 2 above)
Comment (Austinpayne) [reply Sources/SQLiteNIO/SQLiteDataConvertible.swift:135]: I was unfortunately not able to implement this one. The problem is that the equality issue manifests because internally `Date` converts unix epoch seconds to a reference date from January 1, 2001 and visa versa. On the encoding side, `Date.sqliteData` uses `Date.timeIntervalSince1970` which adds the time between the unix epoch and the reference date which results in an extra decimal digit in these cases. While this feels disjointed, on the storing side SQLite handles this situation fine. So we are at least better off in the sense that now dates compare equal.

What do you think @gwynne? Since we can't change the to using `Date.timeIntervalSinceReferenceDate` on the encode side without breaking existing users this feels like the best we can do for now.

Comment (gwynne) [reply Sources/SQLiteNIO/SQLiteDataConvertible.swift:135]: Given the exact inputs you specified in your new test, I was able to more directly examine the underlying issue:
```swift
  1> import Foundation
  2> Date.timeIntervalBetween1970AndReferenceDate
$R0: Double = 978307200
  3> Date(timeIntervalSinceReferenceDate: 689658914.293192)
$R1: Foundation.Date = 2022-11-09 03:55:14 UTC
  4> Date(timeIntervalSinceReferenceDate: 689658914.293192).timeIntervalSince1970
$R2: TimeInterval = 1667966114.2931919
  5> Date(timeIntervalSinceReferenceDate: 689658914.293192).timeIntervalSinceReferenceDate
$R3: TimeInterval = 689658914.29319203
  6> Date(timeIntervalSinceReferenceDate: 689658914.2931919).timeIntervalSinceReferenceDate
$R4: TimeInterval = 689658914.29319191
```
Which ultimately comes down to meaning it's a combination of the precision of `Double` and Tanner's choice of which representation to use - and technically, fractional seconds since either reference point isn't a supported representation at all, to add insult to injury - per https://sqlite.org/datatype3.html#date_and_time_datatype, fractional values are expected to be a Julian day representation (and SQLite does its own rounding, adding even more confusion, but in practice the damage is already done before that point). As you said, however, we can't change the representation without breaking existing users, and the rounding SQLite does on incoming values means the encode is okay as it is anyhow, I think.

So, short version, you did the right thing, all good 😅.

**Review by gwynne — APPROVED** [review]
Okay, LGTM!
